### DISCUSSION TITLE
RunChromVar() custom annotation

### DISCUSSION THREAD

**Discussion by wudustan** [stuart-lab/signac#436] [Q&A] [answered]
Hi, 

Would it be possible to pass the function annotation for promoter regions of genes (as a Granges for example), rather than Jaspar Motifs? I would like to get single-cell level deviation scores for promoter regions for genes of interest in different programmes of expression. Is this possible within Signac currently?

**Comment by timoast** [comment, marked as answer]
You can use the `motif.matrix` parameter in `RunChromVAR()`, and provide a binary peak x class matrix (1-dimensional), where each entry in the column is a 1 if the peak is in a promoter, and 0 otherwise

**Reply by wudustan** [reply]
Hi, could you provide a dummy example of the object in question? Maybe just 3 or so entries so I can understand how it's supposed to be structured?

Many thanks.

**Reply by wudustan** [reply]
@timoast This is my matrix object:

```
> head(motif.matrix, 5)
                   [,1]
chr1-181155-181671    1
chr1-191229-191736    1
chr1-629049-630193    1
chr1-631528-631904    1
chr1-632080-632913    1
```

When I run:

```
> combined
 An object of class Seurat 
 192690 features across 7579 samples within 2 assays 
 Active assay: peaks (172799 features, 172799 variable features)
  1 other assay present: RNA
  2 dimensional reductions calculated: lsi, umap

combined <- RunChromVAR(
  object = combined,
  motif.matrix = motif.matrix,
  genome = BSgenome.Hsapiens.UCSC.hg38
)
```

I get this output:

```
 Computing GC bias per region
 Selecting background regions
 Computing deviations from background
 Constructing chromVAR assay
 Error in CreateAssayObject(data = chromvar.z) : 
  No feature names (rownames) names present in the input matrix
```

I assume my motif.matrix is wrong?

**Reply by timoast** [reply]
You need to add column names to the matrix, eg `colnames(motif.matrix) <- "promoter"`

**Reply by wudustan** [reply]
Thanks @timoast, I have tried that and I have this resulting object, I don't think it's quite right since it only has one feature "promoter" which I assume comes from the `colnames()`:

```
> combined[["promoter"]]
Assay data with 1 features for 7579 cells
First 1 feature:
 promoter 
```

Is the `motif.matrix` supposed to have a column for each 'feature' being the name of a gene promoter region and rows as peak names (eg chr1-12345-67890) then a 1 for each peak in that region?

**Reply by timoast** [reply]
That's what I understood your original goal to be: summarise the promoter accessibility across all genes. Whatever columns are in the motif matrix will become the features in the resulting chromVAR assay, so if you want something different you just need to adjust the motif matrix accordingly.

**Reply by wudustan** [reply]
Ah OK. When you said initially to use "a binary peak x class matrix (1-dimensional)" I assumed it was only supposed to have one column (dimension). If it's a general peaks x features 1/0 matrix, I can make that.

**Reply by wudustan** [reply]
Hi, @timoast I used a sparse matrix where rows were the features, columns were gene promoter names, and 1 where there was an intersect. 

It ran and I am getting an output back in the `<seurat_object>@assays$<assay_name>@data` but I am also getting this warning when chromVAR runs:

```
Warning message:
In UseMethod("depth") :
  no applicable method for 'depth' applied to an object of class "NULL"
```

Is this something to worry about?

**Reply by summer-tao** [reply]
@wudustan  May I ask how you generate the motif.matrix, please?

**Reply by wudustan** [reply]
@summer-tao Here is some code I worked through to annotate promoter regions for some genes of interest. I'm not saying it is 100% correct @timoast can advise if there is something that needs to be changed. I'm figuring this out (making it up) as I go along as I'm quite new to scATAC too

```
library(TxDb.Hsapiens.UCSC.hg38.knownGene)
txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene

## import gene promoters from TxDB
gene.prom <- ensembldb::promoters(txdb, upstream = 1500, downstream = 500)

# keep my regular chromosomes
gene.prom <- gene.prom[seqnames(gene.prom) %in% paste0("chr", c(1:22, "X", "Y"))]
# reorder levels
seqlevels(gene.prom) <- paste0("chr", c(1:22, "X", "Y"))

# remove versioning from transcript name (the number after the full stop)
gene.prom$tx_name <- gsub("\\.[0-9]+", "", gene.prom$tx_name)

## get genes by transcript ID
ens.mart <- useEnsembl(biomart = "genes", dataset = "hsapiens_gene_ensembl")
promoter.bm <- getBM(attributes = c("ensembl_gene_id", "ensembl_transcript_id", "hgnc_symbol", "gene_biotype"),
                     mart = ens.mart, 
                     filters = "ensembl_transcript_id",
                     values = gene.prom$tx_name)

# keep only coding genes and remove some duplicates
promoter.bm <- promoter.bm[promoter.bm$gene_biotype=="protein_coding", ]
promoter.bm <- promoter.bm[!duplicated(promoter.bm$ensembl_transcript_id), ]

# attach the info to the metadata
elementMetadata(gene.prom) <- cbind(elementMetadata(gene.prom), 
                                    promoter.bm[match(gene.prom$tx_name, 
                                                      promoter.bm$ensembl_transcript_id), ])
# strip NA names
gene.prom <- gene.prom[!is.na(gene.prom$ensembl_transcript_id)]
# some hugo names are empty but not NA
gene.prom <- gene.prom[!gene.prom$hgnc_symbol==""]

# hacky way to merge regions where there are multiple non-contiguous ranges per gene
gene.prom <- unlist(reduce(split(gene.prom, gene.prom$hgnc_symbol), min.gapwidth = 1e7))
gene.prom <- gene.prom[!duplicated(names(gene.prom))]

##
unq.genes <- c("your list of unique gene names to annotate")
##

# subset only genes we're interested in because ChromVAR takes a long time
gene.prom <- gene.prom[names(gene.prom) %in% unq.genes]

# make an empty matrix, fitted to our data and make it sparse by default, and make a holding column
motif.matrix <- matrix(0, ncol = 1, nrow = length(rownames(combined@assays$peaks)))
rownames(motif.matrix) <- rownames(combined@assays$peaks)

motif.matrix <- as(motif.matrix, "CsparseMatrix")
colnames(motif.matrix) <- "idx"

my.seurat.object # object of class Seurat generated by following the vignettes from Signac

# fill the holder column, and bind it to our matrix
# put a 1 in each feature that is in our promoters, 0 otherwise, per gene

# this is hacky and slow - hence why I suggest only annotating a gene list
# I'm not used to working with sparse matrices so this can definitely be optimised

for(i in 1:length(gene.prom)){
  peaks.overlap <- my.seurat.object@assays$peaks@ranges[subjectHits(findOverlaps(gene.prom[i], 
                                                                                 my.seurat.object@assays$peaks@ranges, 
                                                                                 maxgap = 10000, 
                                                                                 ignore.strand = TRUE))]
  if (length(peaks.overlap) > 0){
    tmp.mat <- motif.matrix[, "idx", drop = FALSE]
    tmp.mat[paste0(peaks.overlap@seqnames, "-", as.character(peaks.overlap@ranges)), ] <- 1
    colnames(tmp.mat) <- names(gene.prom)[i]
    motif.matrix <- cbind(tmp.mat, motif.matrix)
  }
  
  if ( i %% 1000 == 0) message ("Done ", i, "\n")
}

# strip holding column
motif.matrix <- motif.matrix[, !colnames(motif.matrix)=="idx"]
motif.matrix <- motif.matrix[, order(colnames(motif.matrix), decreasing = FALSE)]

## run chromVAR
combined <- RunChromVAR(object = my.seurat.object,
                        motif.matrix = motif.matrix,
                        genome = BSgenome.Hsapiens.UCSC.hg38,
                        new.assay.name = "promoter")
```

**Reply by summer-tao** [reply]
Super super super thank you. You are sooooo nice. Really thanks for your sharing. Wish you all the best wishes!

**Comment by summer-tao** [comment]
@wudustan @timoast  Dear Tim, dear wudustan,

I have found another way that can calculate the promoter activity and easier to visualize all genes. I am willing to share with you. And @timoast , could you please have a look and check whether this is correct?

So the idea is:  Firstly I generate the "bodyPromoterRNA" assay, which counts the gene body and upstream 2000b as the gene activity. And then I generate another "bodyRNA" assay, which counts only the gene body. And finally, I get the "promoter" assay by "bodyPromoterRNA" minus "bodyRNA". Here I attach the codes:


####creat bodyPromoterRNA assay####

gene.activities <- GeneActivity(
  combined,
  assay = "ATAC",
  extend.upstream = 2000,
  extend.downstream = 200,
  verbose = TRUE)
#check whether there are duplicated genes
rownames(gene.activities)[duplicated(rownames(gene.activities))]
#remove th eduplicated
gene.activities <- gene.activities[rownames(gene.activities)!="Csnk2a1",]
# add the gene activity matrix to the Seurat object as a new assay
combined[['bodyPromoterRNA']] <- CreateAssayObject(counts = gene.activities)
combined<- NormalizeData(
  object = combined,
  assay = 'bodyPromoterRNA',
  normalization.method = 'LogNormalize',
  scale.factor = median(combined$nCount_bodyPromoterRNA)
)

###########generate bodyRNA assay
gene.activitiesBody <- GeneActivity(
  combined,
  assay = "ATAC",
  **extend.upstream = 0,
  extend.downstream = 0,**
  verbose = TRUE)
#check whether there are duplicated genes
rownames(gene.activitiesBody)[duplicated(rownames(gene.activitiesBody))]
#remove th eduplicated
gene.activitiesBody <- gene.activitiesBody[rownames(gene.activitiesBody)!="Csnk2a1",]
# add the gene activity matrix to the Seurat object as a new assay
combined[['bodyRNA']] <- CreateAssayObject(counts = gene.activitiesBody)
combined<- NormalizeData(
  object = combined,
  assay = 'bodyRNA',
  normalization.method = 'LogNormalize',
  scale.factor = median(combined$nCount_bodyRNA)
)

############calculate
bodyPromoter <- combined@assays$bodyPromoterRNA
body <- Tcombined@assays$bodyRNA

promoteri <- bodyPromoter@counts@i - body@counts@i
View(promoteri)
promoterP <- bodyPromoter@counts@p - body@counts@p

###########generate promoterRNA assay
combined[['promoterRNA']] <-combined@assays$bodyRNA
combined@assays$promoterRNA@counts@i <- promoteri
combined@assays$promoterRNA@counts@p <- promoterP
########isolate the promoterAssay and save
promoterAssay <- combined@assays$promoterRNA
saveRDS(promoterAssay ,"promoterAssay.rds")

combined@assays$promoterRNA <- promoterAssay
DefaultAssay(combined) <- "promoterRNA"


**So, now we can visualize the promoter activity of every gene using featureplot, vlnplot, dorplot and so on, same as gene activity.**
Really hope this makes sense.

Best,
Tao.
